[Bug]: hamcrest required in class path in 2.55.0

### What happened?

Found during 2.55.0RC1 release validation against Dataflow template

This used to work in Beam 2.54.0

```java

// Main.java
PipelineOptions options = PipelineOptionsFactory.fromArgs(argv).withValidation().create();
```

```gradle

// Build.gradle

dependencies {
    implementation(group: 'org.apache.beam', name: 'beam-sdks-java-io-google-cloud-platform', version: "$beam_version") {
        exclude group: "org.hamcrest", module: "hamcrest"
    }
    implementation(group: 'org.apache.beam', name: 'beam-sdks-java-core', version: "$beam_version") {
        exclude group: "org.hamcrest", module: "hamcrest"
    }
    implementation(group: 'org.apache.beam', name: 'beam-runners-direct-java', version: "$beam_version") {
        exclude group: "org.hamcrest", module: "hamcrest"
    }
}
```

However, in Beam 2.55.0RC1, there is exception thrown (see https://github.com/GoogleCloudPlatform/DataflowTemplates/actions/runs/8251849699?pr=1361):

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/hamcrest/Matcher
        ...
        at org.apache.beam.sdk.options.PipelineOptionsFactory.<clinit>(PipelineOptionsFactory.java:543)
        at com.github.abacn.MinimumStreaming.main(MinimumStreaming.java:19)
Caused by: java.lang.ClassNotFoundException: org.hamcrest.Matcher
        at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        ... 34 more
```

The interface failed to create is TestBigQueryOptions which extends TestPipelineOptions. In TestPipelineOptions, some methods return org.hamcrest.Matcher: https://github.com/apache/beam/blob/986f67fd92b27dee0c95c5965de98ba6f84a0143/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineOptions.java#L45

It seems beam actually always requires org.hamcrest present in runtime class path. Indeed, another minimum test fails on both 2.54.0 and 2.55.0:

```java

PipelineOptions options = PipelineOptionsFactory.fromArgs(argv).withValidation().as(TestPipelineOptions.class);
```

```gradle

dependencies {
    implementation(group: 'org.apache.beam', name: 'beam-sdks-java-core', version: "$beam_version") {
        exclude group: "org.hamcrest", module: "hamcrest"
    }
    implementation(group: 'org.apache.beam', name: 'beam-runners-direct-java', version: "$beam_version") {
        exclude group: "org.hamcrest", module: "hamcrest"
    }
}
```


### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/google-cloud-platform/build.gradle
+++ b/sdks/java/io/google-cloud-platform/build.gradle
@@ -116,8 +116,8 @@ dependencies {
   permitUnusedDeclared library.java.grpc_google_cloud_pubsublite_v1 // BEAM-11761
   implementation library.java.guava
   implementation library.java.http_client
-  provided library.java.hamcrest
-  permitUnusedDeclared library.java.hamcrest // BEAM-11761
+  implementation library.java.hamcrest
+  permitUnusedDeclared library.java.hamcrest // Needed for PipelineOptionRegistrar auto-service that registered TestPipelineOptions interfaces
   implementation library.java.http_core
   implementation library.java.jackson_core
   implementation library.java.jackson_databind

